tests/analysis/impl/dense/gpcca_test.py

```diff
@@ -0,0 +1,630 @@
+import pytest
+import numpy as np
+
+from scipy.linalg import hilbert, pinv
+from scipy.sparse import csr_matrix, issparse
+from typing import Optional
+
+from tests.get_input import get_known_input, mu
+from tests.numeric import assert_allclose
+from tests.conftest import skip_if_no_petsc_slepc
+from msmtools.analysis.dense.gpcca import (
+    GPCCA,
+    gpcca_coarsegrain,
+    _do_schur,
+    _fill_matrix,
+    _gram_schmidt_mod,
+    _indexsearch,
+    _initialize_rot_matrix,
+    _objective,
+    _opt_soft,
+    _cluster_by_isa,
+)
+
+eps = np.finfo(np.float64).eps * 1e10
+
+
+def _assert_schur(
+    P: np.ndarray, X: np.ndarray, RR: np.ndarray, N: Optional[int] = None
+):
+    if N is not None:
+        np.testing.assert_array_equal(P.shape, [N, N])
+        np.testing.assert_array_equal(X.shape, [N, N])
+        np.testing.assert_array_equal(RR.shape, [N, N])
+
+    assert np.all(np.abs(X @ RR - P @ X) < eps), np.abs(X @ RR - P @ X)
+    assert np.all(np.abs(X[:, 0] - 1) < eps), np.abs(X[:, 0])
```

Comment:
@michalk8 , need to be careful here. By default, we don't compute the whole Schur decomposition but only a partial one, which means this will fail.

Reply:
This can only pass if X is square. To check partial schur decompositions, we need to use something else.